applications/machine_learning/src/modules/ei_data_forwarder.c

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <zephyr.h>
+#include <stdio.h>
+#include <drivers/uart.h>
+
+#include "sensor_event.h"
+
+#define MODULE ei_data_forwarder
+#include "module_state_event.h"
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(MODULE, CONFIG_ML_APP_EI_DATA_FORWARDER_LOG_LEVEL);
+
+#define UART_LABEL		CONFIG_ML_APP_EI_DATA_FORWARDER_UART_DEV
+#define UART_BUF_SIZE		CONFIG_ML_APP_EI_DATA_FORWARDER_BUF_SIZE
+#define DATA_FORMAT		"%.2f"
+
+static const struct device *dev;
+static atomic_t uart_busy;
+
+enum state {
+	STATE_DISABLED,
+	STATE_ACTIVE,
+	STATE_ERROR
+};
+
+static enum state state;
+
+
+static void report_error(void)
+{
+	state = STATE_ERROR;
+	module_set_state(MODULE_STATE_ERROR);
+}
+
+static int snprintf_error_check(int res, size_t buf_size)
+{
+	if ((res < 0) || (res >= buf_size)) {
+		LOG_ERR("snprintf returned %d", res);
+		report_error();
+		return res;
+	}
+
+	return 0;
+}
+
+static bool handle_sensor_event(const struct sensor_event *event)
+{
+	if (state != STATE_ACTIVE) {
+		return false;
+	}
+
+	__ASSERT_NO_MSG(sensor_event_get_data_cnt(event) > 0);
+
+	/* Ensure that previous sensor_event was sent. */
+	if (!atomic_cas(&uart_busy, false, true)) {
+		LOG_ERR("UART not ready");
+		LOG_ERR("Sampling frequency is too high");
+		report_error();
+		return false;
+	}
+
+	static uint8_t buf[UART_BUF_SIZE];
+	const float *data_ptr = sensor_event_get_data_ptr(event);
+	int pos = 0;
+	int tmp = snprintf(&buf[pos], sizeof(buf) - pos, DATA_FORMAT, *data_ptr);
```

Comment:
please put that in the look
you will check against data_size==0 plus avoid code duplication

Reply:
if this is about the comma maybe the following would do
```
for (i = 0; i < 2*size-1; i++) {
   if (i != 0)
      tmp = sprintf(......, ",");
   else
      tmp = sprintf(...., DATA_FORMAT, &data[i/2]);

   error_check(tmp);
}
```

Reply:
Changed